### PR TITLE
Feat/time offsets (WIP)

### PR DESCRIPTION
**Pull request by dbuezas**

### CODE REVIEW

**Comment by dbuezas** [comment]
@zanna-37, now that you have separated all unrelated fixes, I rebased my offsets branch.

If you have time to close the duplicated datapoints issue, please PR to this branch :). Otherwise, I hope to it sooner or later

**Comment by dbuezas** [comment]
* I hope i can get on it sooner or later

**Comment by dbuezas** [comment]
@zanna-37 I rebased now, which includes the better management of the "fake datapoints" and (inspired by you) reworked the git history into "do all the code" and "add a line to the readme" 😁 
With that, I think it is almost done 🎉 

I think the only thing missing for the offsets is to rework the `removeOutsideRange` function in the `cache.ts` file, so it accounts for the offsets. It should work fine now, but it handles the cache very inefficiently (it will delete stuff just to refetch it again an instant later).

Another option is to be less strict removing stuff from the cache. Originally this functionality was there only to ensure that plotly autorange would put in the screen the right xaxis range to respect hours_to_show. Now that's not necessary anymore.

File: src/plotly-graph-card.ts [hunk 1]
```diff
@@ -464,8 +472,7 @@ export class PlotlyGraph extends HTMLElement {
     const was = this.parsed_config;
     this.parsed_config = newConfig;
     const is = this.parsed_config;
-    if (!this.contentEl) return;
```
Comment (zanna-37) [inline src/plotly-graph-card.ts:467]: I think this was meant to stay.

File: src/plotly-graph-card.ts [hunk 2]
```diff
@@ -44,10 +43,14 @@ function patchLonelyDatapoints(xs: Datum[], ys: Datum[]) {
...
-function extendLastDatapointToPresent(xs: Datum[], ys: Datum[]) {
+function extendLastDatapointToPresent(
+  xs: Datum[],
+  ys: Datum[],
+  offset: number
+) {
   if (xs.length === 0) return;
   const last = JSON.parse(JSON.stringify(ys[ys.length - 1]));
-  xs.push(new Date());
+  xs.push(new Date(Date.now() + offset));
```
Comment (zanna-37) [inline src/plotly-graph-card.ts:53]: `extendLastDatapointToPresent(...)` should be disabled when an offset is set. Those traces reflects values that are not related to the present.

**Review by zanna-37 — CHANGES_REQUESTED** [review]

**Review by zanna-37 — CHANGES_REQUESTED** [review]
https://github.com/dbuezas/lovelace-plotly-graph-card/pull/109/files#diff-c0352c6cdab9f9d6e1251e51657ccfc84873e2c7b361622d9aa4094c04c11093R171-R176

Offset should be considered here.

File: src/plotly-graph-card.ts (hunk 1 above)
Comment (dbuezas) [reply src/plotly-graph-card.ts:467]: It's not needed anymore:)
Now that the HTML is setup in the constructor, this is always defined

File: src/plotly-graph-card.ts (hunk 2 above)
Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: Let's discuss this.
An extreme case is a monthly statistic, without extending, it will look like the plot just stops at end of last month.
IMO, of the behavior is to extend to the present, then it should be consistent or it will surprise the user.
I'll put some example screenshots later

**Comment by dbuezas** [comment]
> Offset should be considered on `src/plotly-graph-card.ts`, rows `171-176`.
> 
> https://github.com/dbuezas/lovelace-plotly-graph-card/pull/109/files#diff-c0352c6cdab9f9d6e1251e51657ccfc84873e2c7b361622d9aa4094c04c11093R171-R176

The cache doesn't store an offset history. The offset is applied in the getter in `cache.ts/getHistory(Entity)`

File: src/plotly-graph-card.ts (hunk 2 above)
Comment (zanna-37) [reply src/plotly-graph-card.ts:53]: Well, it depends. 😂
When using offsets, we are talking about things that occurred in the past and is now being shifted.
If we talk about forecast it doesn't make sense to extend, because the forecast is referred to a specific point in time, extending it is nonsensical.
However, when talking a tank level, if the level stays at X liters, and it doesn't change, it would make sense to extend it.

Neither is perfect, but maybe the second scenario is more common so we should stick with your vision. Opinions?

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: True. Give it a try when you test it. I think when you see how a part of the plot disappears as soon as the offset is added you may get the same "this is wrong" feeling.

It is true that it can be seen as a (poor) forecast, but it can also be interpreted as "the last known state at that point".

Let me know what you find out from experimenting. We can always add an option to deactivate the extension

Comment (FrnchFrgg) [reply src/plotly-graph-card.ts:53]: For temperatures and statistics (which are currently timed at the start of their period) extending is surprising, because it extends horizontally between one and two hours but the real temperature didn't stay constant. It looks wierd. If you extended by adding the current state instead of the last fetched one it would look better, but that's IMHO too smart, non generalizable and out of scope.

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: Good point. For statistics, extending is particularly bad indeed, irrespective of offsets. (e.g #129).
I think this should be configurable then. 
I'd appreciate opinions on these:
* should `extendToPresent` default to true or false? I'm changing my opinion back and forth.
* from this thread, it seems to me that the pros and cons of extending apply to non offset entities as much as offset ones, and should therefore be considered independent and almost unrelated.

I'm not sure what you mean with:

> extended by adding the current state instead of the last fetched

You mean like what happens with `refresh_interval: auto`?

Comment (FrnchFrgg) [reply src/plotly-graph-card.ts:53]: I tend to think `extendToPresent` should default to `false`. In fact, I think it should not exist at all. Extending the data from the last valid value to the end of the configured time range is worthwhile when showing "instantaneous" data like the historical one and thus could default to `true` in that case. Indeed, the sensor actually did keep that state until the end of the time range, or else some `unavailable`, or another value, would have been present.

Note that when using offsets, then on purpose you do not fetch some data that lies between the end of the shown graph and the current time, so extending to the present time *is* harmful because you are going to present false data. The rule IMHO is to extend to the end of the visible timespan, that is to the right of the graph, but a) only if it is prior to the current time b) it should *not* change the actual visible timespan nor ignore any real data that would be in the "end of configured fetch range to current time" zone.

And for data where each point represents a certain time range on its own, then extending it does not work. AFAICT statistics give a new point for new ranges even if the value didn't change so they do not work like the normal history data and thus do not need this workaround anyway. Also, I think HA does not send a "fake" data point in the case of statistics (that fake data point is probably designed exactly to behave like a "extend to the end of the fetched range", BTW).

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: All good points. 
BTW, the fake datapoint is at the start of the range, I didn't see it at the end (or maybe it adds it if the range ends in the past? If that is the case, I have a bug in the code to fix)

I am starting to see how extending to the present (as opposed to the max of the range ) is indeed an issue when offsets come into play. So I think I agree fully with this.

The issue I have with extending by default except if there is an offset, is that there is going to be a jump when you go from offset 0 to anything else.

I'll make a version with all of this configurable so we can play a bit to see if there is a sweet spot.

Maybe something like this:
* default is to extend states (so equal to the standard graph card of HA)
* but not for statistics
* extension is not to the present, but to the limit of the fetched range

I think seeing it in action will make the right solution be evident. I'll code it soon

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: > `extendLastDatapointToPresent(...)` should be disabled when an offset is set. Those traces reflects values that are not related to the present.

@zanna-37 I'm starting to understand why you said this. With offsets, the fetched range doesn't necessarily have anything to do with the present.

Comment (FrnchFrgg) [reply src/plotly-graph-card.ts:53]: The fake datapoint at the beginning of the range is probably to "bring into the range" the last datapoint that occured just before the range, so that the left of the graph looks as if it was a cropped version of a bigger graph. I thought that HA did so on the end of the range too, but I may be mistaken.

The "extendLastDatapointToPresent" is not a problem if the offset is moving the trace later. I suppose that what you really want is to extend the last datapoint to the time `now + offset`.

The idea is that when you offset, it is just a translation of the trace by the vector `(offset, 0)`. And the "fake point" of `(now, last_val)` that you add without offset creates a segment from `(last_time, last_val)` to `(now, last_val)` which gets translated into a segment from `(last_time + offset, last_val)` to `(now + offset, last_val)`.

I am not sure if you should crop that `now + offset` to the right end of the graph (to avoid an automatic zoom extending the range in the future initially) or if having data past `now` is not a problem. When the user pans to the right the extension should appear without cropping if offset is positive, though.

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: > The "extendLastDatapointToPresent" is not a problem if the offset is moving the trace later.

If one offsets right (positive) by a month and the visible range is a week, the only the section [now - 1month -1 week, now - 1month] will he fetched] . Extending that to now will just add a huge 1 month straight line (visible when you zoom).
If one offsets left (negative), then it is less problematic

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: > if having data past now is not a problem

X axis scaling is now done explicitly, so this is less of an issue now

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: > Note that my proposition makes having an offset (say a small enough offset that you cannot see, but non-zero nonetheless) totally transparent and not changing suddenly the way the trace looks.

I think i got you now. Whatever we do, it should be equivalent to expanding the true last state first and the offseting. (Not offsetting and then expanding the last fetched one, which may not be the latest state of the sensor)

Comment (FrnchFrgg) [reply src/plotly-graph-card.ts:53]: > I think i got you now. Whatever we do, it should be equivalent to expanding the true last state first and the offseting.

If there is an existing state right of the visible range, it should probably be fetched for continuity purposes. Extending the last visible state to the right only works when plotting the values piecewise constant. With other interpolation methods the result is wrong.

If there is no such later state, you may want to assume the state did not change by extending to the right, but never past `now+offset`. I believe this should default to true for normal history and false for statistics, but should be configurable by the user.

Sadly, I don't think HA has a suitable API to fetch « whatever is the next point after this time » so the first paragraph may be infeasible at least for sensors updating very irregularly.

Comment (dbuezas) [reply src/plotly-graph-card.ts:53]: > never past now+offset

absolutly

---

>  I believe this should default to true for normal history

Agree

---

> and false for statistics

---

It took me a second to realise why, but I agree now too

---

> but should be configurable by the user.

Also agree. Luckily very easy to implement with the new better structured code.

---

> I don't think HA has a suitable API to fetch « whatever is the next point after this time »

Yeah, I didn't find such an API either. 

---

I'll give this a try now and finally merge this PR

**Comment by dbuezas** [comment]
Config for extending works.

image

```yaml
type: custom:plotly-graph-dev
entities:
  - entity: sensor.electricity_total
    extend_to_present: false
    line:
      width: 10
  - entity: sensor.electricity_total
    extend_to_present: true 
hours_to_show: 0.03
update_interval: 1
```

I think this PR can be merged.
@zanna-37 Should I merge or wait? If you still intend to review the readme of offsets let me know! :)

**Comment by zanna-37** [comment]
Yes, I'll write the README for offset, maybe tomorrow morning.

Regarding the last code changes, ff you want to wait I can give a look at the code.

**Comment by dbuezas** [comment]
> Yes, I'll write the README for offset, maybe tomorrow morning.

Thanks :) !!!

> Regarding the last code changes, ff you want to wait I can give a look at the code.

Will be very appreciated! I removed the cache clearing code (*), but had to still add it before plotting to ensure the yaxis auto range applies to the visible section. (So if you reset, the data won't be deleted from cache, but will still not be plotted)

(*) It's kind of hard when combined with offsets, as the same entity may be shown with different offsets in the same plot.

**Comment by zanna-37** [comment]
> It's kind of hard when combined with offsets, as the same entity may be shown with different offsets in the same plot.

It's difficult for me to review the code since I'm losing track of what methods works on the actual values of the offsetted ones.

By the way, it might be a good idea to use `entityname+offset` as the cache key.

**Comment by dbuezas** [comment]
> it might be a good idea to use entityname+offset

The idea is to not fetch the same values twice. The entity key determines the raw data (more or less), and the offsets are applied only before plotting. Multiple traces can be derived from the same cached entry (each with different offsets, lambdas, etc). That's why the cache key only holds the entity name, attribute and statistic type and period.

**Comment by dbuezas** [comment]
> what methods works on the actual values of (or?) the offsetted ones.

The offset is only applied to decide what to fetch, and then right before plotting. I think i may remove the offset logic from the cache to make that more evident

**Comment by dbuezas** [comment]
In any case, I'm fairly confident about the offset logic, so the readme would be the most valuable contribution :)

**Comment by dbuezas** [comment]
But you can point out at something specific that is confusing, I'll gladly try to improve it :)

**Comment by zanna-37** [comment]
Either we keep the cache as points that will be plotted (offset already applied) and in this case we keep separate traces for the same entity with different offset.
OR
We keep the raw entity in the cache but we must be careful about merging and deleting the out-of-range.

> so the readme would be the most valuable contribution

I know, it's coming.

> point out at something specific

it's just documentation, for methods and vars, but don't worry, I know it's a burden and the problem is mine, not yours 😂

**Comment by dbuezas** [comment]
> OR
We keep the raw entity in the cache but we must be careful about merging and deleting the out-of-range.

100%. That's the approach i took in the PR. To solve the deleting problem, i just stopped deleting the cache, but filter the "post processed" data before giving it to plotlyjs. 

I just saw that there is a bug (plotlyjs again), where the autorrange button (or double clicking) autorranges to the widest past scales instead of the visible data. It is probably storing that as a side effect somewhere, instead of traversing the data again. I'll try to work around that next.

**Comment by zanna-37** [comment]
> I just stopped deleting the cache

So when the graph stays there for too long we are filling the memory indefinitely. Right?

**Comment by dbuezas** [comment]
> > I just stopped deleting the cache
> 
> So when the graph stays there for too long we are filling the memory indefinitely. Right?

Yes, that's the caveat. And slower operations (filtering happens linearly, no binary search for the boundaries or anything of the sort.
I don't think it should be too tragic though. Worse case we could clear the cache every hour.
Maybe i refactor the cache eventually and implement it better. At the moment, the same info will ve fetched more than once if you plot multiple attributes of the same entity

File: readme.md [hunk 3]
```diff
@@ -247,6 +247,35 @@ entities:
...
+```yaml
+type: custom:plotly-graph
+offset: 6h
+entities:
+  - entity: sensor.weather_24h_forecast
+    offset: -1d
+  - entity: sensor.actual_temperature
+```
+
+### Extend_to_present
```
Comment (dbuezas) [inline readme.md:261]: @FrnchFrgg 
Here's the toggle for `extend_to_present` with the defaults you suggested

**Comment by zanna-37** [comment]
README ready at #142.

File: readme.md [hunk 4]
```diff
@@ -247,6 +247,35 @@ entities:
...
+The boolean `extend_to_present` will take the last known datapoint and "expand" it to the present by creating a duplicate and setting its date to `now`.
+This is useful to make the plot look fuller.
+It's recommended to turn it off when using `offset`s, or when setting the mode of the trace to `markers`.
+Defaults to `true` for state history, and `false` for statistics.
+
+```yaml
+type: custom:plotly-graph
+entities:
+  - entity: sensor.weather_24h_forecast
+    mode: "markers"
```
[the remark this answers is not in the document]
Comment (dbuezas) [reply readme.md:291]: Not within this repo's readme, it's a plotlyjs feature, so it's in their docs

File: src/plotly-graph-card.ts [hunk 5]
```diff
@@ -432,6 +456,7 @@ export class PlotlyGraph extends HTMLElement {
             throw new Error(
               `period: "${entity.period}" is not valid. Use ${STATISTIC_PERIODS}`
             );
+          entity.extend_to_present ??= !!entity.statistic;
```
Comment (zanna-37) [inline src/plotly-graph-card.ts:459]: This condition should be flipped. Now `extend_to_present` is true when using statistics.

Comment (dbuezas) [reply src/plotly-graph-card.ts:459]: Woho, good catch

**Comment by dbuezas** [comment]
@zanna-37 fyi: in v3, the offsets are called `time_offset`. I had to rename the property to avoid a collision with a plotlyjs config to offset bar charts.
